### PR TITLE
Custom eq for ResourceKey

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]

File: provider/fs/src/export/fs_exporter.rs [hunk 1]
```diff
@@ -133,7 +133,7 @@ impl DataExporter for FilesystemExporter {
                 .expect("present iff file.1 is present")
                 .lock()
                 .expect("poison")
-                .push((key, options.clone(), format!("{:x}", hash.finalize())));
+                .push(format!("{}/{}: {:x}", key, options, hash.finalize()));
```
Comment (sffc) [inline provider/fs/src/export/fs_exporter.rs:136]: Praise: This is better and will always produce a stable ordering going forward

File: provider/datagen/src/transform/cldr/plurals/mod.rs [hunk 2]
```diff
@@ -25,24 +25,24 @@ impl From<&SourceData> for PluralsProvider {
 
 impl PluralsProvider {
     fn get_rules_for(&self, key: ResourceKey) -> Result<&cldr_serde::plurals::Rules, DataError> {
-        match key {
-            CardinalV1Marker::KEY => self
-                .source
+        if key == CardinalV1Marker::KEY {
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/plurals/mod.rs:28]: Nit: Why did you change this function? I think `match` on ResourceKey should still work (just less efficient)

File: provider/datagen/src/transform/cldr/list/mod.rs [hunk 3]
```diff
@@ -43,11 +43,14 @@ impl<M: ResourceMarker<Yokeable = ListFormatterPatternsV1<'static>>> ResourcePro
             .expect("CLDR file contains the expected language")
             .list_patterns;
 
-        let (wide, short, narrow) = match M::KEY {
-            AndListV1Marker::KEY => (&data.standard, &data.standard_short, &data.standard_narrow),
-            OrListV1Marker::KEY => (&data.or, &data.or_short, &data.or_narrow),
-            UnitListV1Marker::KEY => (&data.unit, &data.unit_short, &data.unit_narrow),
-            _ => return Err(DataError::custom("Unknown key for ListFormatterPatternsV1")),
+        let (wide, short, narrow) = if M::KEY == AndListV1Marker::KEY {
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/list/mod.rs:46]: Nit, here and below: can't we keep using `match`?

**Review by sffc — COMMENTED** [review]
I support this change! Just a few comments

File: provider/datagen/src/transform/cldr/plurals/mod.rs (hunk 2 above)
Comment (robertbastian) [reply provider/datagen/src/transform/cldr/plurals/mod.rs:28]: Nope, you can only match on `const` values that have a derived `PartialEq`.

Comment (robertbastian) [reply provider/datagen/src/transform/cldr/plurals/mod.rs:28]: This is why I didn't want to add fields to `ResourceKey`

File: provider/core/src/resource.rs [hunk 4]
```diff
@@ -110,30 +110,30 @@ unsafe impl EqULE for ResourceKeyHash {}
...
     // in a compiled binary.
     path: &'static str,
     hash: ResourceKeyHash,
 }
 
-// I don't think this lint makes sense, see https://github.com/rust-lang/rust-clippy/issues/2627.
-// It's not possible to violate the hash invariant (k1 == k2 -> hash(k1) == hash(k2)) without
-// using non-determinism or unsafe code when PartialEq is derived.
-#[allow(clippy::derive_hash_xor_eq)]
+impl PartialEq for ResourceKey {
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply provider/core/src/resource.rs:121]: These are related because changing the databake match to not match on `ResourceKey` lets us remove the derived `PartialEq` and instead do something more efficient.

Should we compare (hash, string)? Then we get a quick exit if the hashes don't match.

File: provider/core/src/datagen/mod.rs [hunk 5]
```diff
@@ -72,15 +77,19 @@ macro_rules! make_exportable_provider {
 
         impl $crate::datagen::DataConverter<$crate::buf::BufferMarker, $crate::datagen::HeapStatsMarker> for $provider {
             fn convert(&self, key: $crate::ResourceKey, from: DataPayload<$crate::buf::BufferMarker>) -> Result<$crate::DataPayload<$crate::datagen::HeapStatsMarker>, $crate::datagen::ReturnedPayloadError<$crate::buf::BufferMarker>> {
-                use $crate::datagen::ReturnedPayloadError;
-                match key {
+                #![allow(non_upper_case_globals)]
+                // Reusing the struct names as identifiers
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply provider/core/src/datagen/mod.rs:81]: It's really annoying that we have to create this one-time use constants :(

[the remark this answers is not in the document]
Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:81]: could use some kind of macro to do `force_const!(...)`

[the remark this answers is not in the document]
Comment (robertbastian) [reply provider/core/src/datagen/mod.rs:81]: Blocks don't work as patterns if that's what you mean.

[the remark this answers is not in the document]
Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:81]: yeah. I was hoping `{ .. }` const syntax would work but that's only in generics

File: provider/core/src/resource.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/resource.rs:121]: ok.

[the remark this answers is not in the document]
Comment (sffc) [reply provider/core/src/resource.rs:121]: > Should we compare (hash, string)? Then we get a quick exit if the hashes don't match.

That would address my concern

**Review by sffc — COMMENTED** [review]
Please check the path string in `impl PartialEq`, and consider recovering the `match`

File: provider/datagen/src/transform/cldr/plurals/mod.rs (hunk 2 above)
Comment (robertbastian) [reply provider/datagen/src/transform/cldr/plurals/mod.rs:28]: I considered that but the if/else is more concise

**Review by sffc — COMMENTED** [review]
LGTM and merged
